language_service/src/signature_help.rs

```diff
@@ -0,0 +1,156 @@
+    signature_help: Option<SignatureHelp>,
+    display: CodeDisplay<'a>,
+}
+
+impl<'a> Visitor<'a> for SignatureHelpFinder<'a> {
+    fn visit_item(&mut self, item: &'a ast::Item) {
+        if span_contains(item.span, self.offset) {
+            walk_item(self, item);
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a ast::Expr) {
+        if span_contains(expr.span, self.offset) {
+            match &*expr.kind {
+                ast::ExprKind::Call(callee, args) if span_contains(args.span, self.offset) => {
+                    walk_expr(self, args);
+                    if self.signature_help.is_none() {
+                        let callee = unwrap_parens(callee);
+                        if let ast::ExprKind::Path(path) = &*callee.kind {
+                            if let Some(resolve::Res::Item(item_id)) =
+                                self.compilation.unit.ast.names.get(path.id)
+                            {
+                                if let (Some(item), _) = find_item(self.compilation, item_id) {
+                                    if let qsc::hir::ItemKind::Callable(callee) = &item.kind {
+                                        // Check that the callee has parameters to give help for
+                                        if !matches!(&callee.input.kind, hir::PatKind::Tuple(items) if items.is_empty())
+                                        {
+                                            let sig_info = SignatureInformation {
+                                                label: self
+                                                    .display
+                                                    .hir_callable_decl(callee)
+                                                    .to_string(),
+                                                documentation: None,
+                                                parameters: self.get_params(callee),
+                                            };
+
+                                            self.signature_help = Some(SignatureHelp {
+                                                signatures: vec![sig_info],
+                                                active_signature: 0,
+                                                active_parameter: process_args(args, self.offset),
+                                            });
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+                _ => walk_expr(self, expr),
+            }
+        }
+    }
+}
+
+fn unwrap_parens(expr: &ast::Expr) -> &ast::Expr {
+    match &*expr.kind {
+        ast::ExprKind::Paren(inner) => unwrap_parens(inner),
+        _ => expr,
+    }
+}
+
+impl SignatureHelpFinder<'_> {
+    fn get_params(&self, decl: &hir::CallableDecl) -> Vec<ParameterInformation> {
+        let offset = self.display.get_param_offset(decl);
```

Comment:
Actually it'd be really helpful if you could add a comment on top that describes what we're doing with a code example. Such as:

```
// operation Foo(bar: Int, baz: Int)
//               ^param 1 start
//                         ^parameter 2 start
```

Feel free to modify my suggestion per your taste. I find that my eyes kind of glaze over all the Rust code but a visual example really gets across what the function is trying to accomplish.

Reply:
I left a similar example as a comment along with a small description of what the function is doing.